keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Geometric.java

```diff
@@ -0,0 +1,54 @@
+package io.improbable.keanu.distributions.discrete;
+
+import io.improbable.keanu.distributions.DiscreteDistribution;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.tensor.intgr.IntegerTensor;
+import io.improbable.keanu.vertices.dbl.KeanuRandom;
+
+public class Geometric implements DiscreteDistribution {
+
+    private final DoubleTensor p;
+
+    public static DiscreteDistribution withParameters(DoubleTensor p) {
+        return new Geometric(p);
+    }
+
+    private Geometric(DoubleTensor p) {
+        this.p = p;
+    }
+
+    @Override
+    public IntegerTensor sample(long[] shape, KeanuRandom random) {
+        DoubleTensor numerator = random.nextDouble(shape).logInPlace();
+        DoubleTensor denominator = p.unaryMinus().plusInPlace(1.0).logInPlace();
+
+        return numerator.divInPlace(denominator).floorInPlace().toInteger().plusInPlace(1);
```

Comment:
isn't the `floor` redundant when followed by `toInteger`?

Reply:
I gave it a quick go by inverting the CDF, and got something slightly different. I expect I've made a mistake somewhere but would be good to find out where. I got `log(1-r)/log(1-p)`.

Reply:
is it useful to compare it to `Exponential#sample`?

    return random.nextDouble(shape).logInPlace().timesInPlace(lambda).unaryMinusInPlace();

Reply:
re: Floor being redundant - I think *technically* it is, but that is only true because we will hopefully only ever generate non-negative numbers.  If you were dealing with negative numbers I think our behaviour might be inconsistent between Scalar and ND4J tensors (with scalars not being consistent with .floor()

Reply:
re: the sampling logic - I've done my working from scratch again, and chatted it through with Wes and I'm pretty happy with what we have given the tests I've written look good.

Certainly from manual verification, if you generate a tonne of samples - you get a halving of the number of such samples each time the k value increases.

Reply:
👍